--- ../../samtools-1.10/htslib-1.10/Makefile	2019-11-26 10:01:13.000000000 -0600
+++ vendor/samtools-1.10/htslib-1.10/Makefile	2019-12-19 14:28:10.000000000 -0600
@@ -27,7 +27,8 @@
 RANLIB = ranlib
 
 # Default libraries to link if configure is not used
-htslib_default_libs = -lz -lm -lbz2 -llzma -lcurl
+#htslib_default_libs = -lz -lm -lbz2 -llzma -lcurl
+htslib_default_libs = -lz -lm -lbz2 -llzma
 
 CPPFLAGS =
 # TODO: probably update cram code to make it compile cleanly with -Wc++-compat
@@ -179,7 +180,7 @@
 # Without configure we wish to have a rich set of default figures,
 # but we still need conditional inclusion as we wish to still
 # support ./configure --disable-blah.
-NONCONFIGURE_OBJS = hfile_libcurl.o
+#NONCONFIGURE_OBJS = hfile_libcurl.o
 
 PLUGIN_EXT  =
 PLUGIN_OBJS =
@@ -219,7 +220,7 @@
 	echo '#define HAVE_LZMA_H 1' >> $@
 	echo '#endif' >> $@
 	echo '#define HAVE_DRAND48 1' >> $@
-	echo '#define HAVE_LIBCURL 1' >> $@
+#	echo '#define HAVE_LIBCURL 1' >> $@
 
 # And similarly for htslib.pc.tmp ("pkg-config template").  No dependency
 # on htslib.pc.in listed, as if that file is newer the usual way to regenerate
